src/ddsrt/src/events/kqueue/events_kqueue.c

```diff
@@ -0,0 +1,388 @@
+ * to an internal pipe.
+ */
+struct ddsrt_event_queue
+{
+  queued_event_t*         events;   /**< container for stored events*/
+  size_t                  nevents;  /**< number of events stored*/
+  size_t                  cevents;  /**< capacity of events stored*/
+  size_t                  ievents;  /**< current iterator for getting the next triggered event*/
+  size_t                  ndeleted; /**< number of events monitored by kevent deleted since last call to wait function*/
+  queued_event_t*         newevents;  /**< container for modified events*/
+  size_t                  nnewevents; /**< number of new events added since last call to wait function*/
+  size_t                  cnewevents; /**< capacity for new events*/
+  int                     kq;       /**< kevent polling instance*/
+  struct kevent*          kevents;  /**< array which kevent uses to write back to, has identical size as this->events*/
+  ddsrt_mutex_t           lock;     /**< for keeping adds/deletes from occurring simultaneously */
+  int                     interrupt[2]; /**< pipe for interrupting waits*/
+};
+
+/**
+* @brief Initializes an event queue.
+*
+* Will set the counters to 0 and create the containers for triggers and kevents.
+* Will create a kevent kernel event instance and open the interrupt pipe.
+*
+* @param[in,out] queue The queue to initialize.
+*
+* @retval DDS_RETCODE_OK
+*             The queue was initialized succesfully.
+* @retval DDS_RETCODE_ERROR
+*             - There was an issue with reserving memory for the (k)event queue.
+*             - Kevent instance or interrupt pipe could not be initialized correctly.
+*/
+static dds_return_t ddsrt_event_queue_init(ddsrt_event_queue_t* queue) ddsrt_nonnull_all;
+
+dds_return_t ddsrt_event_queue_init(ddsrt_event_queue_t* queue)
+{
+  queue->nevents = 0;
+  queue->cevents = EVENTS_CONTAINER_DELTA;
+  queue->ievents = 0;
+  queue->ndeleted = 0;
+  queue->nnewevents = 0;
+  queue->cnewevents = EVENTS_CONTAINER_DELTA;
+
+  /*create kevents array*/
+  queue->kevents = ddsrt_malloc(sizeof(struct kevent) * queue->cevents);
+  if (NULL == queue->kevents)
+    goto alloc_fail_0;
+  queue->events = ddsrt_malloc(sizeof(queued_event_t) * queue->cevents);
+  if (NULL == queue->events)
+    goto alloc_fail_1;
+  queue->newevents = ddsrt_malloc(sizeof(queued_event_t) * queue->cnewevents);
+  if (NULL == queue->newevents)
+    goto alloc_fail_2;
+
+  /*create kevent polling instance */
+  if (-1 == (queue->kq = kqueue()))
+    goto kq_fail;
+  else if (-1 == fcntl(queue->kq, F_SETFD, fcntl(queue->kq, F_GETFD) | FD_CLOEXEC))
+    goto pipe0_fail;
+  /*create interrupt pipe */
+  else if (-1 == pipe(queue->interrupt))
+    goto pipe0_fail;
+  else if (-1 == fcntl(queue->interrupt[0], F_SETFD, fcntl(queue->interrupt[0], F_GETFD) | FD_CLOEXEC) ||
+           -1 == fcntl(queue->interrupt[1], F_SETFD, fcntl(queue->interrupt[1], F_GETFD) | FD_CLOEXEC))
+    goto pipe1_fail;
+
```

Comment:
I think one should also set "non-blocking" on the interrupt pipe. The risk is minimal, but without making it non-blocking, the `write` call in `signal` could block.

Reply:
This has now been implemented.